File: sdk-python/tests/integ/mixed_wait_flow.py
# Portions of this file are derived from indeedeng/iwf-java-sdk.
# Those portions are licensed under the Apache License, Version 2.0.
# See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
#
# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications are licensed under the Super Durable Source License 1.0.
# Third-Party Materials remain under the Apache License, Version 2.0.
# See LICENSE and LEGACY_NOTICES.md.

from datetime import timedelta

from dex import (
    Context,
    Flow,
    Step,
    StepDecision,
    StepList,
    StepOptions,
    Timer,
    Wait,
    go_to,
    graceful_complete,
)


class MixedTimerStep(Step[int]):
    def __init__(self, options: StepOptions) -> None:
        self.options = options

    def wait_for(self, context: Context, input: int) -> Wait:
        del context, input
        return Wait.all_of(Timer.by_duration(timedelta(seconds=1)))

    def execute(self, context: Context, input: int) -> StepDecision:
        del context
        return graceful_complete(input + 1)

    def get_step_options(self) -> StepOptions:
        return self.options


class MixedImmediateStep(Step[int]):
    def __init__(self, second: MixedTimerStep, options: StepOptions) -> None:
        self.second = second
        self.options = options

    def execute(self, context: Context, input: int) -> StepDecision:
        del context
        return go_to(self.second, input + 1)

    def get_step_options(self) -> StepOptions:
        return self.options


class MixedWaitFlow(Flow[int]):
    def __init__(self) -> None:
        shared = StepOptions(execute_method_timeout=timedelta(seconds=5))
        self.second = MixedTimerStep(shared)
        self.first = MixedImmediateStep(self.second, shared)

    def get_steps(self) -> StepList[int]:
        return StepList.start_step(self.first).other_steps(self.second)

File: sdk-typescript/test/integ/mixed_wait_flow.ts
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

import {
  StepList,
  Timer,
  Wait,
  doubleCodec,
  goTo,
  gracefulComplete,
  type Context,
  type Flow,
  type Step,
  type StepDecision,
  type StepOptions,
} from "../../src/index.js";

class MixedTimerStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(private readonly options: StepOptions) {}

  public getStepType(): string {
    return "MixedTimerStep";
  }

  public waitFor(_context: Context, _input: number): Wait {
    return Wait.until(Timer.byDuration(1_000));
  }

  public execute(_context: Context, input: number): StepDecision {
    return gracefulComplete(input + 1);
  }

  public getStepOptions(): StepOptions {
    return this.options;
  }
}

class MixedImmediateStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(
    private readonly second: MixedTimerStep,
    private readonly options: StepOptions,
  ) {}

  public getStepType(): string {
    return "MixedImmediateStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return goTo(this.second, input + 1);
  }

  public getStepOptions(): StepOptions {
    return this.options;
  }
}

export class MixedWaitFlow implements Flow<number> {
  private readonly shared: StepOptions = { executeMethodTimeoutMs: 5_000 };
  private readonly second = new MixedTimerStep(this.shared);
  private readonly first = new MixedImmediateStep(this.second, this.shared);

  public getFlowType(): string {
    return "MixedWaitFlow";
  }

  public getSteps() {
    return StepList.startStep(this.first).otherSteps(this.second);
  }
}
